felix/bpf-gpl/parsing.h

```diff
@@ -63,6 +64,23 @@ static CALI_BPF_INLINE int parse_packet_ip(struct cali_tc_ctx *ctx) {
 		}
 	}
 
+	// For ingress packet, check if it has local route matched, for unknown packet,
+	// just allow it with CALI_REASON_RT_UNKNOWN
+	if (CALI_F_FROM_HEP) {
```

Comment:
I'd take this out of parsing, probably to https://github.com/projectcalico/calico/blob/master/felix/bpf-gpl/tc.c#L505

My reasoning would be that in most setups we do not need to spend time on the route lookup for every packet that comes from a HEP. We would like to do it after we miss in conntrack. Also the benefit would be that we create a conntrack entry and ever subsequent packet would take the same path. Also this would be subject to a policy. Not sure whether that is what you want, but it feels like it should be :thinking:

Reply:
> I think it's not so desired to do the route lookup after we miss in conntrack. For the incoming packet from HEP(physical interfaces I think), we need to do an initial check for the incoming packet if its dest is for our Calico system by check if its dst addr is route known...

We definitely do not want to do route lookup for every packet on HEP. In most cases that would be followed by conntrack lookup if that is for something within the cluster. We want to do it the other way around. If there is a miss, we classify the flow and then create conntrack so that we have the same classification for subsequent traffic. The difference here is performance and not what ultimately happens to that packet.

Reply:
> So here we also should not check the unknown route traffic against FIB by bpf_fib_lookup (in forward_or_drop()), since in some systems, the lookup result would be successful like this:
-0 [088] d.s. 1810775.267240: bpf_trace_printk: enp9s0---I: Traffic is towards the host namespace, doing Linux FIB lookup
-0 [088] d.s. 1810775.267243: bpf_trace_printk: enp9s0---I: FIB lookup succeeded - with neigh
-0 [088] d.s. 1810775.267244: bpf_trace_printk: enp9s0---I: Got Linux FIB hit, redirecting to iface 2.
-0 [088] d.s. 1810775.267245: bpf_trace_printk: enp9s0---I: Traffic is towards host namespace, marking with 0x3000000.
-0 [088] d.s. 1810775.267247: bpf_trace_printk: enp9s0---I: Final result=ALLOW (0). Program execution time: 31307ns
-0 [088] d.s. 1810775.267249: bpf_trace_printk: enp9s0---E: New packet at ifindex=2; mark=3000000
-0 [088] d.s. 1810775.267250: bpf_trace_printk: enp9s0---E: Final result=ALLOW (3). Bypass mark bit set.
>
> and it's a wrong processing here since for the packet of a mark of 3000000 at the egress direction would be discarded by the system.

Why is "packet of a mark of 3000000 at the egress direction" discarded by the system? That mark says, don't look at it, just allow it:

>  4944
  4945           <idle>-0       [088] d.s. 1810775.267249: bpf_trace_printk: enp9s0---E: New packet at ifindex=2; mark=3000000
  4946
  4947           <idle>-0       [088] d.s. 1810775.267250: bpf_trace_printk: enp9s0---E: Final result=ALLOW (3). Bypass mark bit set.

Reply:
> > I think it's not so desired to do the route lookup after we miss in conntrack. For the incoming packet from HEP(physical interfaces I think), we need to do an initial check for the incoming packet if its dest is for our Calico system by check if its dst addr is route known...
> 
> We definitely do not want to do route lookup for every packet on HEP. In most cases that would be followed by conntrack lookup if that is for something within the cluster. We want to do it the other way around. If there is a miss, we classify the flow and then create conntrack so that we have the same classification for subsequent traffic. The difference here is performance and not what ultimately happens to that packet.

Hi Tomas:
Thanks for your comment.
For the issue of the route lookup for every packet on HEP, I think what you concerned here is performance. It's just adding a bpf map(cali_v4_routes) lookup action here which I think should be very quick and have very little effect on the final performance since there are so many actions remained.

On the other side, I think what Cilium had done about this issue:
There is a function:
handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
	    __u32 ipcache_srcid __maybe_unused, const bool from_host)
in https://github.com/cilium/cilium/blob/master/bpf/bpf_host.c#L460
which calls an endpoint lookup action for every incoming packet :
/* Lookup IPv4 address in list of local endpoints and host IPs */
	ep = lookup_ip4_endpoint(ip4);
in https://github.com/cilium/cilium/blob/master/bpf/bpf_host.c#L565,
This lookup action is also done agains a map name ENDPOINTS_MAP, for which I think it's sth like our route map here.
and if there is not valid ep (ep is Nil) found here, it would let it go:
	/* Below remainder is only relevant when traffic is pushed via cilium_host.
	 * For traffic coming from external, we're done here.
	 */
	if (!from_host)
		return CTX_ACT_OK;
in https://github.com/cilium/cilium/blob/master/bpf/bpf_host.c#L577, here !from_host means the packet comes from net devices(NETDEV), which is the same meaning for Calico's from HEP.
This handle_ipv4() function would be called by 
static __always_inline int
tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_host)
in https://github.com/cilium/cilium/blob/master/bpf/bpf_host.c#L656
then called by:
__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV4_FROM_NETDEV)
int tail_handle_ipv4_from_netdev(struct __ctx_buff *ctx)
{
	return tail_handle_ipv4(ctx, 0, false);
}
in https://github.com/cilium/cilium/blob/master/bpf/bpf_host.c#L683
Here the "false" means the packet comes from the netdev.

So I think Cilium also has similar processing for unknown traffic, and AFAIK, Cilium ebpf performance is also very good.

Reply:
I cannot quite comment on the cilium code as I am not familiar with it. However, if we can have the same feature with less overhead, why not to do that. Otherwise one look up here and one look up there add up :sweat_smile: 

Would you be able to test whether the patch a propose would work for you? :pray: